crates/stages/src/stage.rs

```diff
@@ -140,14 +140,14 @@ pub trait Stage<DB: Database>: Send + Sync {
     /// Execute the stage.
     async fn execute(
         &mut self,
-        tx: &mut Transaction<'_, DB>,
+        provider: &mut DatabaseProviderRW<'_, DB>,
```

Comment:
We should rename ShareableDatabase to `ProviderFactory` or something. And I don't _love_ calling the type `DatabaseProvider` as it wraps a TX, maybe `TxProvider` and a `TxProviderFactory`?

Reply:
> We should rename ShareableDatabase to ProviderFactory or something.

 Yeah, I agree

>maybe TxProvider and a TxProviderFactory?

The way I see it, `DatabaseProvider`, provides data from the database... I'm open to change it, but I feel like `TxProvider` will make things even more confusing, given the amount of transaction types that exist already.